diff --git a/hdk_setup.sh b/hdk_setup.sh
index e3d8482..d58f3ec 100644
--- a/hdk_setup.sh
+++ b/hdk_setup.sh
@@ -153,8 +153,11 @@ do
   if [ ! -e $hdk_shell_dir ]; then
       mkdir -p $hdk_shell_dir || { err_msg "Failed to create $hdk_shell_dir"; return 2; }
   fi
+
   # Use curl instead of AWS CLI so that credentials aren't required.
-  curl -s https://s3.amazonaws.com/$s3_shell_dir/$shell_file.sha256 -o $hdk_file.sha256 || { err_msg "Failed to download HDK shell's $shell_file version from $s3_shell_dir/$shell_file.sha256 -o $hdk_file.sha256"; return 2; }
+  curl -s https://s3.amazonaws.com/$s3_shell_dir/$shell_file.sha256 -o $hdk_file.sha256 # || { err_msg "Failed to download HDK shell's $shell_file version from $s3_shell_dir/$shell_file.sha256 -o $hdk_file.sha256"; return 2; }
+  sleep 1
+
   if grep -q '<?xml version' $hdk_file.sha256; then
     err_msg "Failed to download HDK shell's $shell_file version from $s3_shell_dir/$shell_file.sha256"
     cat $hdk_file.sha256
